src/plonk/permutation/prover.rs

```diff
@@ -166,76 +217,125 @@ impl<C: CurveAffine> Committed<C> {
         impl Iterator<Item = Polynomial<C::Scalar, ExtendedLagrangeCoeff>> + 'a,
     ) {
         let domain = &pk.vk.domain;
+        let chunk_len = pk.vk.cs.degree() - 2;
+
+        let constructed = Constructed {
+            sets: self
+                .sets
+                .iter()
+                .map(|set| ConstructedSet {
+                    permutation_product_poly: set.permutation_product_poly.clone(),
+                    permutation_product_blind: set.permutation_product_blind,
+                })
+                .collect(),
+        };
+
         let expressions = iter::empty()
-            // l_0(X) * (1 - z(X)) = 0
-            .chain(Some(
-                Polynomial::one_minus(self.permutation_product_coset.clone()) * &pk.l0,
-            ))
-            // z(X) \prod (p(X) + \beta s_i(X) + \gamma) - z(omega^{-1} X) \prod (p(X) + \delta^i \beta X + \gamma)
-            .chain(Some({
-                let mut left = self.permutation_product_coset.clone();
-                for (values, permutation) in p
-                    .columns
+            // Enforce only for the first set.
+            // l_0(X) * (1 - z_0(X)) = 0
+            .chain(self.sets.first().map(|first_set| {
+                Polynomial::one_minus(first_set.permutation_product_coset.clone()) * &pk.l0
+            }))
+            // Enforce only for the last set.
+            // l_last(X) * (z_l(X)^2 - z_l(X)) = 0
+            .chain(self.sets.last().map(|last_set| {
+                ((last_set.permutation_product_coset.clone() * &last_set.permutation_product_coset)
+                    - &last_set.permutation_product_coset)
+                    * &pk.l_last
+            }))
+            // Except for the first set, enforce.
+            // l_0(X) * (z_i(X) - z_{i-1}(\omega^(last) X)) = 0
+            .chain(
+                self.sets
                     .iter()
-                    .map(|&column| match column.column_type() {
-                        Any::Advice => {
-                            &advice_cosets[pk.vk.cs.get_any_query_index(column, Rotation::cur())]
-                        }
-                        Any::Fixed => {
-                            &fixed_cosets[pk.vk.cs.get_any_query_index(column, Rotation::cur())]
-                        }
-                        Any::Instance => {
-                            &instance_cosets[pk.vk.cs.get_any_query_index(column, Rotation::cur())]
-                        }
+                    .skip(1)
+                    .zip(self.sets.iter())
+                    .map(|(set, last_set)| {
+                        (
+                            set.permutation_product_coset.clone(),
+                            last_set.permutation_product_coset_last.clone().unwrap(),
```

Comment:
Isn't it simpler to do `(set.permutation_product_coset.clone() - last_set.permutation_product_coset_last.clone().unwrap()) * &pk.l0` here, rather than two maps?

Reply:
Yeah, I was trying to avoid returning an iterator that needed ownership or a reference to `self.sets` which is being dropped at the end of this method. It's more efficient to just map like you're saying and just collect it, not sure why I didn't just do that.